### PR TITLE
fix: secrets generation, more bootstrap tests

### PR DESCRIPTION
**Pull request by Morriz**
also removed the deps on relative paths for mounting env dir, so we can run and debug locally

### CODE REVIEW

**Review by j-zimnowoda — CHANGES_REQUESTED** [review]
The very first job is failing and the second one is able to bootstrap correctly

first job:
```

2021-12-13T12:32:55.826Z otomi:hfValues:info No teams or cluster info found. ENV_DIR is potentially empty.
2021-12-13T12:32:55.827Z otomi:bootstrap:info Intalling artifacts from otomi/core:v0.14.54
2021-12-13T12:32:55.829Z otomi:bootstrap:info Copied bin files
2021-12-13T12:32:55.839Z otomi:bootstrap:info Copied vscode folder
2021-12-13T12:32:55.961Z otomi:bootstrap:info Stored loose YAML schema at: /home/app/stack/env/.vscode/values-schema.yaml
2021-12-13T12:32:55.963Z otomi:bootstrap:log Copying skeleton files
2021-12-13T12:32:56.009Z otomi:bootstrap:log Loading chart values from /secret/values.yaml
2021-12-13T12:32:56.009Z otomi:bootstrap:info Checking if secret/default/otomi-generated-passwords already exists
2021-12-13T12:32:56.204Z otomi:values:writeValuesToFile:info Values were written to /home/app/stack/env/env/settings.yaml
2021-12-13T12:32:56.205Z otomi:values:writeValuesToFile:info No changes for /home/app/stack/env/env/secrets.settings.yaml, skipping...
2021-12-13T12:32:56.206Z otomi:values:writeValuesToFile:info Values were written to /home/app/stack/env/env/cluster.yaml
2021-12-13T12:32:56.207Z otomi:values:writeValuesToFile:info No changes for /home/app/stack/env/env/charts/cert-manager.yaml, skipping...
2021-12-13T12:32:56.208Z otomi:values:writeValuesToFile:info No changes for /home/app/stack/env/env/charts/external-dns.yaml, skipping...
2021-12-13T12:32:56.211Z otomi:values:writeValues:info All values were written to ENV_DIR
2021-12-13T12:32:56.211Z otomi:generateSecrets:info Extracting secrets
2021-12-13T12:32:56.214Z otomi:generateSecrets:info First round of templating
2021-12-13T12:32:56.228Z otomi:generateSecrets:info Parsing values for second round of templating
2021-12-13T12:32:56.228Z otomi:generateSecrets:info Gather all values for the second round of templating
2021-12-13T12:32:56.229Z otomi:generateSecrets:info Second round of templating
2021-12-13T12:32:56.324Z otomi:generateSecrets:info Generated all secrets
2021-12-13T12:32:56.326Z otomi:createCustomCA:info Generating custom root CA
2021-12-13T12:32:56.597Z otomi:createCustomCA:info Generated CA key pair
2021-12-13T12:32:56.606Z otomi:values:writeValuesToFile:info No changes for /home/app/stack/env/env/settings.yaml, skipping...
2021-12-13T12:32:56.607Z otomi:values:writeValuesToFile:info No changes for /home/app/stack/env/env/secrets.settings.yaml, skipping...
2021-12-13T12:32:56.607Z otomi:values:writeValuesToFile:info No changes for /home/app/stack/env/env/cluster.yaml, skipping...
2021-12-13T12:32:56.608Z otomi:values:writeValuesToFile:info No changes for /home/app/stack/env/env/charts/cert-manager.yaml, skipping...
2021-12-13T12:32:56.608Z otomi:values:writeValuesToFile:info No changes for /home/app/stack/env/env/charts/external-dns.yaml, skipping...
2021-12-13T12:32:56.609Z otomi:values:writeValues:info All values were written to ENV_DIR
2021-12-13T12:32:57.052Z otomi:global:error Error: in /home/app/stack/helmfile.tpl/helmfile-dump-all.yaml: error during ../helmfile.d/snippets/derived.gotmpl.part.0 parsing: template: stringTemplate:73:63: executing "stringTemplate" at <$v.otomi.adminPassword>: map has no entry for key "adminPassword"
    at hfCore (/home/app/stack/dist/src/common/hf.js:46:30)
    exit code: 1
    at ChildProcess.<anonymous> (/home/app/stack/node_modules/zx/dist/bundle.cjs:15343:22)
    at ChildProcess.emit (node:events:390:28)
    at maybeClose (node:internal/child_process:1064:16)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:301:5)
```

second job
```
2021-12-13T12:33:00.176Z otomi:hfValues:info No teams or cluster info found. ENV_DIR is potentially empty.
2021-12-13T12:33:00.177Z otomi:bootstrap:info Intalling artifacts from otomi/core:v0.14.54
2021-12-13T12:33:00.179Z otomi:bootstrap:info Copied bin files
2021-12-13T12:33:00.190Z otomi:bootstrap:info Copied vscode folder
2021-12-13T12:33:00.295Z otomi:bootstrap:info Stored loose YAML schema at: /home/app/stack/env/.vscode/values-schema.yaml
2021-12-13T12:33:00.297Z otomi:bootstrap:log Copying skeleton files
2021-12-13T12:33:00.329Z otomi:bootstrap:log Loading chart values from /secret/values.yaml
2021-12-13T12:33:00.330Z otomi:bootstrap:info Checking if secret/default/otomi-generated-passwords already exists
2021-12-13T12:33:00.426Z otomi:bootstrap:info Found secret/default/otomi-generated-passwords secrets on cluster, recovering
2021-12-13T12:33:00.521Z otomi:values:writeValuesToFile:info Values were written to /home/app/stack/env/env/settings.yaml
2021-12-13T12:33:00.522Z otomi:values:writeValuesToFile:info No changes for /home/app/stack/env/env/secrets.settings.yaml, skipping...
2021-12-13T12:33:00.522Z otomi:values:writeValuesToFile:info Values were written to /home/app/stack/env/env/cluster.yaml
2021-12-13T12:33:00.523Z otomi:values:writeValuesToFile:info No changes for /home/app/stack/env/env/charts/cert-manager.yaml, skipping...
2021-12-13T12:33:00.523Z otomi:values:writeValuesToFile:info No changes for /home/app/stack/env/env/charts/external-dns.yaml, skipping...
2021-12-13T12:33:00.526Z otomi:values:writeValues:info All values were written to ENV_DIR
2021-12-13T12:33:00.526Z otomi:generateSecrets:info Extracting secrets
2021-12-13T12:33:00.529Z otomi:generateSecrets:info First round of templating
2021-12-13T12:33:00.541Z otomi:generateSecrets:info Parsing values for second round of templating
2021-12-13T12:33:00.542Z otomi:generateSecrets:info Gather all values for the second round of templating
2021-12-13T12:33:00.543Z otomi:generateSecrets:info Second round of templating
2021-12-13T12:33:00.639Z otomi:generateSecrets:info Generated all secrets
2021-12-13T12:33:00.641Z otomi:bootstrap:info Skipping custom RootCA generation
2021-12-13T12:33:00.649Z otomi:values:writeValuesToFile:info No changes for /home/app/stack/env/env/settings.yaml, skipping...
2021-12-13T12:33:00.650Z otomi:values:writeValuesToFile:info Values were written to /home/app/stack/env/env/secrets.settings.yaml
2021-12-13T12:33:00.651Z otomi:values:writeValuesToFile:info Values were written to /home/app/stack/env/env/charts/secrets.drone.yaml
2021-12-13T12:33:00.651Z otomi:values:writeValuesToFile:info Values were written to /home/app/stack/env/env/charts/secrets.gitea.yaml
2021-12-13T12:33:00.652Z otomi:values:writeValuesToFile:info Values were written to /home/app/stack/env/env/charts/secrets.harbor.yaml
2021-12-13T12:33:00.653Z otomi:values:writeValuesToFile:info Values were written to /home/app/stack/env/env/charts/secrets.keycloak.yaml
2021-12-13T12:33:00.653Z otomi:values:writeValuesToFile:info Values were written to /home/app/stack/env/env/charts/secrets.kubeapps.yaml
2021-12-13T12:33:00.654Z otomi:values:writeValuesToFile:info Values were written to /home/app/stack/env/env/charts/secrets.loki.yaml
2021-12-13T12:33:00.655Z otomi:values:writeValuesToFile:info Values were written to /home/app/stack/env/env/charts/secrets.oauth2-proxy.yaml
2021-12-13T12:33:00.657Z otomi:values:writeValues:info All values were written to ENV_DIR
2021-12-13T12:33:01.265Z otomi:gen-sops:warn No sops information given. Assuming no sops enc/decryption needed. Be careful!
2021-12-13T12:33:01.266Z otomi:bootstrap:log `otomi.adminPassword` has been generated and is stored in the values repository in `env/secrets.settings.yaml`
2021-12-13T12:33:01.266Z otomi:bootstrap:log You can now use the otomi CLI
2021-12-13T12:33:01.266Z otomi:bootstrap:log Done bootstrapping values
2021-12-13T12:33:01.801Z otomi:prepareEnvironment:info Checking environment
2021-12-13T12:33:01.812Z otomi:apply:info Start apply all
2021-12-13T12:33:02.074Z otomi:apply:log ---

<redacted>
```

File: src/cmd/bootstrap.ts [hunk 1]
```diff
@@ -113,29 +113,31 @@ export const processValues = async (
...
+
+  // do we need to create a custom CA? if so add it to the secrets
+  const cm = get(originalValues, 'charts.cert-manager', {})
+  if (cm.issuer === 'custom-ca' || cm.issuer === undefined) {
+    if (cm.customRootCA && cm.customRootCAKey) {
+      deps.debug.info('Skipping custom RootCA generation')
+    } else generatedSecrets = deps.createCustomCA(originalValues as Record<string, any>)
+  }
+  // we have generated all we need, now store the values and merge in the secrets
+  await deps.writeValues(merge(originalValues, generatedSecrets), true)
```
Comment (j-zimnowoda) [inline src/cmd/bootstrap.ts:140]: why do you want to overwrite values? it can lead to overwriting already generated secrets

File: helmfile.d/snippets/env.gotmpl [hunk 2]
```diff
@@ -1,9 +1,10 @@
...
-{{- $settings := readFile "../env/env/settings.yaml" | fromYaml }}
+{{- $ENV_DIR := env "ENV_DIR" }}
+{{- $teams := readFile (print $ENV_DIR "/env/teams.yaml") | fromYaml }}
+{{- $settings := readFile (print $ENV_DIR "/env/settings.yaml") | fromYaml }}
 {{- $teams := keys $teams.teamConfig.teams}}
-{{- $hasSops := eq (exec "bash" (list "-c" "( test -f ../env/.sops.yaml && echo 'true' ) || echo 'false'")) "true" }}
-{{- $charts := (exec "bash" (list "-c" "find  ../env/env/charts -name '*.yaml' -not -name 'secrets.*.yaml'")) | splitList "\n" }}
-{{- $chartsSecret := (exec "bash" (list "-c" "find  ../env/env/charts -name 'secrets.*.yaml'")) | splitList "\n" }}
+{{- $hasSops := eq (exec "bash" (list "-c" "( test -f $ENV_DIR/.sops.yaml && echo 'true' ) || echo 'false'")) "true" }}
+{{- $charts := (exec "bash" (list "-c" "find  $ENV_DIR/env/charts -name '*.yaml' -not -name 'secrets.*.yaml'")) | splitList "\n" }}
```
Comment (0-sv) [inline helmfile.d/snippets/env.gotmpl:6]: I don't fully understand, when I came up with a proposal to do it like this you told me it wasn't a good idea, so I wonder what has changed?

**Review by 0-sv — APPROVED** [review]

**Review by j-zimnowoda — CHANGES_REQUESTED** [review]
I still observe the issue for the first job:
```
│ 2021-12-14T10:00:09.620Z otomi:global:error Error: in /home/app/stack/helmfile.tpl/helmfile-dump-all.yaml: error during ../helmfile.d/snippets/derived.gotmpl.part.0 parsing: template: stringTemplate:73:63: executing "stringTemplate" at <$v.otomi.adminPassword>: map ha │
│     at hfCore (/home/app/stack/dist/src/common/hf.js:46:30)                                                                                                                                                                                                                  │
│     exit code: 1                                                                                                                                                                                                                                                             │
│     at ChildProcess.<anonymous> (/home/app/stack/node_modules/zx/dist/bundle.cjs:15343:22)                                                                                                                                                                                   │
│     at ChildProcess.emit (node:events:390:28)                                                                                                                                                                                                                                │
│     at maybeClose (node:internal/child_process:1064:16)                                                                                                                                                                                                                      │
│     at Process.ChildProcess._handle.onexit (node:internal/child_process:301:5)                                                                                                                                                                                               │
│
```

File: helmfile.d/snippets/env.gotmpl (hunk 2 above)
Comment (j-zimnowoda) [reply helmfile.d/snippets/env.gotmpl:6]: and I doubt that you have tested it
```
2021-12-14T10:24:44.462Z otomi:global:error Error: in /home/app/stack/helmfile.tpl/helmfile-dump-files.yaml: error during ../helmfile.d/snippets/env.gotmpl.part.0 parsing: template: stringTemplate:2:14: executing "stringTemplate" at <readFile (print $ENV_DIR "/env/teams.yaml")>: error calling readFile: open /env/teams.yaml: no such file or directory
--
5 | at hfCore (/home/app/stack/dist/src/common/hf.js:46:30)
6 | exit code: 1
7 | at ChildProcess.<anonymous> (/home/app/stack/node_modules/zx/dist/bundle.cjs:15343:22)
8 | at ChildProcess.emit (node:events:390:28)
9 | at maybeClose (node:internal/child_process:1064:16)
10 | at Process.ChildProcess._handle.onexit (node:internal/child_process:301:5)
```

Comment (Morriz) [reply helmfile.d/snippets/env.gotmpl:6]: > and I doubt that you have tested it

I installed the latest `secrets-generation` tag and saw no problems. I am betting on the tests to cover edge cases. If we found another one we should add a test for that.

Comment (j-zimnowoda) [reply helmfile.d/snippets/env.gotmpl:6]: It happens during the team deployment.

Comment (Morriz) [reply helmfile.d/snippets/env.gotmpl:6]: it seems otomi.ts is not setting ENV_DIR anymore...it was before I believe

Comment (Morriz) [reply helmfile.d/snippets/env.gotmpl:6]: I see now. We resolve env now late with `env()` and so `otomi.ts` does not see the default set with envalid

Comment (Morriz) [reply helmfile.d/snippets/env.gotmpl:6]: > I don't fully understand, when I came up with a proposal to do it like this you told me it wasn't a good idea, so I wonder what has changed? (referring to using `$ENV_DIR` as opposed to `../env/env..`)

Don't remember exactly. I do know I need it to be able to debug otomi commands targeting the cluster. It will keep working with IN_DOCKER=1 still

File: src/cmd/destroy.ts [hunk 3]
```diff
@@ -32,22 +34,30 @@ const destroyAll = async () => {
     debugStream,
   )
   debug.info('Uninstalling all charts...')
-  await hf({ args: 'destroy' }, { streams: debugStream })
+  await stream(
+    nothrow($`helm ls -a -d --all-namespaces | awk 'NR > 1 { print "-n "$2, $1}' | xargs -L1 helm delete`),
```
Comment (j-zimnowoda) [inline src/cmd/destroy.ts:38]: I believe that idea behind using JS is to leverage its power of parsing, yet I still see parsing SHELL scripts that are quite difficult to understand. 
Would it be possible to parse json output?
E.g.: 
```
data = JSON.parse($`helm ls -a -d --all-namespaces -o json`)
data.foreach( ... )
```

File: chart/otomi/values.yaml [hunk 4]
```diff
@@ -81,3 +81,7 @@ charts:
 #   adminGroupID: ''
 #   # IDP group id used to identify team admin
 #   teamAdminGroupID: ''
+
+# WARNING: setting this to ANY VALUE removes ALL cluster resources
```
Comment (j-zimnowoda) [inline chart/otomi/values.yaml:85]: So this is is interesting use case.
To me otomi chart is like operator. 
If I delete operator shall all resources` created by operator` be deleted?
I would say no by default. In this regard I would suggest to make `destroyUponRemoval: false'
If I do `helm uninstall` the how do I prevent from removing the platform workloads?

**Review by j-zimnowoda — CHANGES_REQUESTED** [review]
Now the issue with adminPassword occurs a bit later in the first job but it it still there ...
```
│ 2021-12-15T08:41:31.926Z otomi:values:writeValues:info All values were written to ENV_DIR                                                                                                                                                                                                                                                                                │
│ 2021-12-15T08:41:32.287Z otomi:createK8sSecret:debug kubectl create secret output:                                                                                                                                                                                                                                                                                       │
│  secret/otomi-generated-passwords created                                                                                                                                                                                                                                                                                                                                │
│ 2021-12-15T08:41:32.387Z otomi:global:error Error: in /home/app/stack/helmfile.tpl/helmfile-dump-all.yaml: error during ../helmfile.d/snippets/derived.gotmpl.part.0 parsing: template: stringTemplate:73:63: executing "stringTemplate" at <$v.otomi.adminPassword>: map has no entry for key "adminPassword"                                                           │
│     at hfCore (/home/app/stack/dist/src/common/hf.js:46:30)                                                                                                                                                                                                                                                                                                              │
│     exit code: 1                                                                                                                                                                                                                                                                                                                                                         │
│     at ChildProcess.<anonymous> (/home/app/stack/node_modules/zx/dist/bundle.cjs:15343:22)                                                                                                                                                                                                                                                                               │
│     at ChildProcess.emit (node:events:390:28)                                                                                                                                                                                                                                                                                                                            │
│     at maybeClose (node:internal/child_process:1064:16)                                                                                                                                                                                                                                                                                                                  │
│     at Process.ChildProcess._handle.onexit (node:internal/child_process:301:5)
```

**Review by j-zimnowoda — CHANGES_REQUESTED** [review]
Unfortunately yet another issue
```
2021-12-15T13:01:40.338Z otomi:global:error Error: in /home/app/stack/helmfile.tpl/helmfile-dump-files.yaml: error during ../helmfile.d/snippets/env.gotmpl.part.0 parsing: template: stringTemplate:6:16: executing "stringTemplate" at <exec "bash" (list "-c" "find  $ENV_DIR/env/charts -name '*.yaml' -not -name 'secrets.*.yaml'")>: error calling exec: command "/usr/bin/bash" exited with non-zero status:
```

There is also issue with default SA account
```
│   name: otomi-critical                                                                                                                                                                                                                                                                                                                                                   │
│ value: 1000000                                                                                                                                                                                                                                                                                                                                                           │
│ 2021-12-15T12:45:07.875Z otomi:global:error Error: Warning: policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+                                                                                                                                                                                                                              │
│ Warning: resource namespaces/default is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically.                                  │
│ Warning: resource serviceaccounts/default is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically.                             │
│ Warning: resource serviceaccounts/default is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically.                             │
│ Warning: resource serviceaccounts/default is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically.                             │
│ Warning: resource serviceaccounts/default is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically.                             │
│ Warning: resource serviceaccounts/default is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically.                             │
│ Warning: resource serviceaccounts/default is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically.                             │
│ Warning: resource serviceaccounts/default is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically.                             │
│ Warning: resource serviceaccounts/default is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically.                             │
│ Warning: resource serviceaccounts/default is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically.                             │
│ Warning: resource serviceaccounts/default is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically.                             │
│ Warning: resource serviceaccounts/default is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically.                             │
│ Warning: resource serviceaccounts/default is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically.                             │
│ Error from server (AlreadyExists): error when creating "/tmp/otomi/deploy-template.yaml": serviceaccounts "default" already exists                                                                                                                                                                                                                                       │
│     at applyAll (/home/app/stack/dist/src/cmd/apply.js:62:23)                                                                                                                                                                                                                                                                                                            │
│     exit code: 1                                                                                                                                                                                                                                                                                                                                                         │
│     at ChildProcess.<anonymous> (/home/app/stack/node_modules/zx/dist/bundle.cjs:15343:22)                                                                                                                                                                                                                                                                               │
│     at ChildProcess.emit (node:events:390:28)                                                                                                                                                                                                                                                                                                                            │
│     at maybeClose (node:internal/child_process:1064:16)                                                                                                                                                                                                                                                                                                                  │
│     at Process.ChildProcess._handle.onexit (node:internal/child_process:301:5)
```

**Review by j-zimnowoda — APPROVED** [review]
